src/object_cmd.cpp

```diff
@@ -120,6 +120,8 @@ void BuildObject(ObjectType type, TileIndex tile, CompanyID owner, Town *town, u
 	assert(o->town != nullptr);
 
 	for (TileIndex t : ta) {
+		if (IsWaterTile(t)) ClearNeighbourNonFloodingStates(t);
```

Comment:
`IsWaterTile` is also excluding coasts. `IsWaterTile` is fine for checking against canals, which can't be placed on half-coast, half-water tiles, but NonFloodingStates can.

Reply:
This is there to re-enable flooding behaviour of neighbouring tiles when replacing a water tile with an object. If it is not re-enabled then the tiles are not flooded when the object is removed.